File: server/src/services/vector-store.ts
import { OpenAI } from 'langchain/llms/openai';
import { RetrievalQAChain, loadQARefineChain } from 'langchain/chains';
import { Chroma } from 'langchain/vectorstores/chroma';
import { OpenAIEmbeddings } from 'langchain/embeddings/openai';
import { StreamingCallbackHandler } from '@/langchain/callbacks/streaming-callback-handler';
import { ConsoleCallbackHandler } from '@/langchain/callbacks/console-callback-handler';
import { loader } from '../loaders';
import { PromptTemplate } from 'langchain/prompts';
import { ChromaClient, Collection } from 'chromadb';
import { processDocumentsIntoObjects } from '@/utils/parse';
import { Document } from 'langchain/dist/document';
import { Where, GetResponse, Metadata } from 'chromadb/dist/main/types';

export interface DocumentsObjectInterface {
  metadata?: Metadata;
  embedding?: Record<string, unknown>;
  document: string;
  id: string;
}

export interface DocumentsInterface {
  ids: string[];
  embeddings?: any[];
  documents: string[];
  metadatas?: any[];
}

export class VectorStore {
  private static store: Chroma | null;

  client: ChromaClient;
  model: OpenAI;
  chatHistory: string[];

  constructor() {
    this.client = new ChromaClient();
    this.model = new OpenAI(
      {
        openAIApiKey: process.env.OPENAI_API_KEY,
        modelName: process.env.DEFAULT_OPENAI_MODEL,
        streaming: true,
        verbose: true,
        temperature: 0.5
      }
      // { organization: process.env.OPENAI_ORG_ID }
    );

    this.chatHistory = [];
  }

  async setCreateChromaStore(name: string) {
    VectorStore.store = new Chroma(
      new OpenAIEmbeddings({ openAIApiKey: process.env.OPENAI_API_KEY }),
      {
        collectionName: name
      }
    );
  }

  async listCollections() {
    const collections = await this.client.listCollections();
    return collections;
  }

  async deleteCollection(name: string) {
    const result = await this.client.deleteCollection({ name });
    return result;
  }

  async getCollection(name: string) {
    return await this.client.getCollection({ name });
  }

  async createCollection(name: string) {
    return await this.client.createCollection({ name });
  }

  async getDocuments(
    collection: Collection,
    query?: object
  ): Promise<Record<string, DocumentsObjectInterface[]>> {
    const documents: GetResponse = await collection.get({
      where: query as Where
    });

    return processDocumentsIntoObjects(documents);
  }

  async addDocuments(file: Express.Multer.File) {
    const { documents, errors } = await loader(file);
    if (VectorStore.store) await VectorStore.store.addDocuments(documents);

    return { documents, errors };
  }

  async deleteDocuments(collectionName: string, filename: string) {
    const collection = await this.client.getCollection({
      name: collectionName
    });

    await collection.delete({ where: { filename } });

    const documents = await collection.get();

    return processDocumentsIntoObjects(documents);
  }

  clearChatHistory() {
    this.chatHistory.length = 0;
  }

  async askQuestion(
    question: string,
    systemPrompt: string,
    queryType: 'simple' | 'refine',
    temperature: number,
    callback: (token: string) => void
  ) {
    console.log(
      'askQuestion',
      question,
      systemPrompt,
      queryType,
      temperature,
      callback
    );
    StreamingCallbackHandler.setStreamCallback(callback);

    const chatPrompt = PromptTemplate.fromTemplate(
      ` ${systemPrompt}
      You are an AI assistant. You will be asked questions about the given documents, you can only use the given documents for information.  You can use your
      memory to help with context or analysis of the documents and to understand the information and question, but you cant make things up. 
      Provide answers in a conversational manner.
      Dont answer with anything like "Based on the provided context" or "Based on Additional Context"

      Answer like a human would.
      If you don't know the answer, don't try to make up an answer 
      Follow the user's instructions carefully. 
      ALWAYS respond using markdown.
     
      Chat History for context:
      {chat_history}

      The Question to be answered: {question}`
    );

    const prompt = await chatPrompt.format({
      question,
      chat_history: [this.chatHistory]
    });

    if (VectorStore.store && queryType === 'refine') {
      const chain = new RetrievalQAChain({
        combineDocumentsChain: loadQARefineChain(this.model),
        retriever: VectorStore.store.asRetriever()
      });

      const res = await chain.call(
        {
          chainType: 'stuff',
          query: prompt,
          temperature
        },
        [new StreamingCallbackHandler(), new ConsoleCallbackHandler()]
      );

      this.chatHistory.push(res.output_text);
    } else {
      if (VectorStore.store) {
        const chain = RetrievalQAChain.fromLLM(
          this.model,
          VectorStore.store.asRetriever()
        );

        const res = await chain.call(
          {
            chainType: 'stuff',
            query: prompt,
            temperature
          },
          [new StreamingCallbackHandler(), new ConsoleCallbackHandler()]
        );

        this.chatHistory.push(res.text);
      }
    }
  }
}

File: pyServer/server/services/vector_store.py
import os
from dotenv import load_dotenv

import chromadb
from chromadb.config import Settings
from chromadb.api.models.Collection import Collection
from chromadb.api.types import GetResult

from langchain.chat_models import ChatOpenAI

from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.vectorstores import Chroma

from langchain.chains import ConversationalRetrievalChain, LLMChain
from langchain.chains.summarize import load_summarize_chain
from langchain.chains.conversational_retrieval.prompts import CONDENSE_QUESTION_PROMPT

from server.services.loaders import LoaderResult

from server.langchain.callbacks.streaming_callback_handler import (
    StreamingCallbackHandler,
)
from server.langchain.callbacks.console_callback_handler import (
    ConsoleCallbackHandler,
)


from server.services.loaders import loader
from server.utils.parse import process_documents_into_objects

from typing import List, Dict, Optional

load_dotenv()


class VectorStore:
    store: Optional[Chroma] = None
    client = chromadb.Client(
        Settings(
            chroma_api_impl="rest",
            chroma_server_host="localhost",
            chroma_server_http_port="8000",
        )
    )

    chain: ConversationalRetrievalChain
    chat_history = []

    @classmethod
    def set_create_chroma_store(cls, name: str):
        VectorStore.store = Chroma(
            embedding_function=OpenAIEmbeddings(
                client="Test", openai_api_key=os.getenv("OPENAI_API_KEY")
            ),
            collection_name=name,
            client=VectorStore.client,
        )

    def list_collections(self):
        return self.client.list_collections()

    def delete_collection(self, name: str):
        self.client.delete_collection(name)

    def get_collection(self, name: str):
        return self.client.get_collection(name)

    def create_collection(self, name: str):
        result = self.client.create_collection(name=name, get_or_create=True)
        return result

    def get_documents(self, collection: Collection, query: Dict = {}):
        documents: GetResult = collection.get(where_document=query)

        return process_documents_into_objects(documents)

    def add_documents(
        self, collection_name: str, file_path: str, filename: str
    ) -> LoaderResult:
        results = loader(file_path, filename)

        print(f"Adding {results} to {results}")

        if not VectorStore.store:
            VectorStore.set_create_chroma_store(collection_name)

        if VectorStore.store and len(results.documents) > 0:
            VectorStore.store.add_documents(results.documents)

        return results

    def delete_documents(self, collection_name: str, filename: str):
        collection = self.get_collection(collection_name)

        print(f"Deleting {filename} from {collection_name}")
        print(f"Collection: {collection}")

        collection.delete(where={"filename": filename})

        documents: GetResult = collection.get()

        return process_documents_into_objects(documents)

    def clear_chat_history(self):
        self.chat_history = []

    def ask_question(
        self,
        question: str,
        system_prompt: str,
        query_type: str,
        temperature: int,
        collection_name: str,
        callback,
    ):
        openai_org_id = os.getenv("OPENAI_ORG_ID")
        openai_api_key = os.getenv("OPENAI_API_KEY")
        default_model = os.getenv("DEFAULT_OPENAI_MODEL") or "gpt-3.5-turbo"

        model_args = {
            "client": "Test",
            "openai_api_key": openai_api_key,
            "model": default_model,
            "callbacks": [StreamingCallbackHandler(), ConsoleCallbackHandler()],
            "streaming": True,
            "verbose": True,
            "temperature": temperature,
        }

        if openai_org_id:
            model_args["openai_organization"] = openai_org_id

        model = ChatOpenAI(**model_args)

        StreamingCallbackHandler.set_stream_callback(callback)

        if not VectorStore.store:
            VectorStore.set_create_chroma_store(collection_name)

        if VectorStore.store:
            if query_type == "refine":
                chain = ConversationalRetrievalChain(
                    question_generator=LLMChain(
                        llm=model, prompt=CONDENSE_QUESTION_PROMPT
                    ),
                    combine_docs_chain=load_summarize_chain(
                        model,
                        chain_type="refine",
                    ),
                    retriever=VectorStore.store.as_retriever(),
                )

            else:
                chain = ConversationalRetrievalChain.from_llm(
                    llm=model,
                    retriever=VectorStore.store.as_retriever(),
                    verbose=True,
                )

            res = chain.run(
                {
                    "question": question,
                    "chat_history": self.chat_history,
                }
            )
            self.chat_history.append((question, res))

        else:
            raise ValueError(
                "Collection is not selected",
            )
